shorebird_code_push/lib/src/shorebird_updater_io.dart

```diff
@@ -0,0 +1,159 @@
+      _isAvailable = false;
+    }
+  }
+
+  late final bool _isAvailable;
+
+  final Updater _updater;
+
+  final IsolateRun _run;
+
+  @override
+  bool get isAvailable => _isAvailable;
+
+  @override
+  Future<Patch?> readCurrentPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.currentPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
+  }
+
+  @override
+  Future<Patch?> readNextPatch() async {
+    if (!_isAvailable) return null;
+
+    return _run(
+      () {
+        try {
+          final patchNumber = _updater.nextPatchNumber();
+          return patchNumber > 0 ? Patch(number: patchNumber) : null;
+        } catch (error) {
+          throw ReadPatchException(message: '$error');
+        }
+      },
+    );
+  }
+
+  @override
+  Future<UpdateStatus> checkForUpdate() async {
+    if (!_isAvailable) return UpdateStatus.unavailable;
+
+    final isUpdateAvailable = await _run(_updater.checkForUpdate);
+    if (isUpdateAvailable) return UpdateStatus.outdated;
+
+    final (current, next) = await (readCurrentPatch(), readNextPatch()).wait;
+    return next != null && current?.number != next.number
+        ? UpdateStatus.restartRequired
+        : UpdateStatus.upToDate;
+  }
+
+  @override
+  Future<void> update() async {
+    if (!_isAvailable) return;
+
+    late final Pointer<UpdateResult> result;
+
+    try {
+      result = await _run(_updater.downloadUpdateWithError);
+
+      if (result == nullptr) {
+        const reason = UpdateFailureReason.unknown;
+        final message = reason.toFailureMessage();
+        throw UpdateException(reason: reason, message: message);
+      }
+
+      final status = result.ref.status;
+
+      if (status == 1) return; // SHOREBIRD_UPDATE_SUCCESS
+
+      final reason = status.toFailureReason();
+      final details = result.ref.message != nullptr
+          ? result.ref.message.cast<Utf8>().toDartString()
+          : 'unknown';
+      final message = reason.toFailureMessage(details);
+      throw UpdateException(message: message, reason: reason);
+    } finally {
+      _updater.freeUpdateResult(result);
+    }
+  }
+}
+
+extension on int {
+  UpdateFailureReason toFailureReason() {
+    switch (this) {
+      case 0: // SHOREBIRD_NO_UPDATE
+        return UpdateFailureReason.noUpdate;
+      case 2: // SHOREBIRD_UPDATE_HAD_ERROR
+        return UpdateFailureReason.downloadFailed;
+      case 3: // SHOREBIRD_UPDATE_IS_BAD_PATCH
+        return UpdateFailureReason.badPatch;
+      case 4: // SHOREBIRD_UPDATE_ERROR
+        return UpdateFailureReason.unknown;
+      default:
+        return UpdateFailureReason.unknown;
+    }
+  }
+}
+
+extension on UpdateFailureReason {
+  String toFailureMessage([String details = '']) {
+    switch (this) {
```

Comment:
We could provide this via a callback into Rust if you wanted. 🤷   Or have rust fill in the message for you?

Reply:
I don't see a huge downside to leaving it in the Dart API since these are errors that are displayed to users consuming the Dart API and I think it makes sense to keep the decoupled (e.g the dart error messages should be more friendly/human-readable whereas the rust layer might not care and an enum is good enough).

Reply:
sgtm.